File: backend/src/models/Subscription.ts
import mongoose, { Schema, Document } from 'mongoose';
import { IClient } from './Client';
import { IServicePlan } from './ServicePlan';
import { IUser } from './User';

export interface ISubscription extends Document {
  clientId: IClient['_id'];
  planId: IServicePlan['_id'];
  ownerId: IUser['_id'];
  status: 'PENDING_ACTIVATION' | 'ACTIVE' | 'PAUSED' | 'CANCELLED' | 'EXPIRED' | 'FAILED';
  startDate: Date;
  nextBillingDate: Date;
  notes?: string;
  paymentFailureCount?: number;
  lastPaymentAttempt?: Date;
}

const SubscriptionSchema: Schema = new Schema({
  clientId: { type: Schema.Types.ObjectId, ref: 'Client', required: true },
  planId: { type: Schema.Types.ObjectId, ref: 'ServicePlan', required: true },
  ownerId: { type: Schema.Types.ObjectId, ref: 'User', required: true },
  status: { type: String, enum: ['PENDING_ACTIVATION', 'ACTIVE', 'PAUSED', 'CANCELLED', 'EXPIRED', 'FAILED'], default: 'PENDING_ACTIVATION' },
  startDate: { type: Date, default: Date.now },
  nextBillingDate: { type: Date, required: true },
  notes: { type: String },
  paymentFailureCount: { type: Number, default: 0 },
  lastPaymentAttempt: { type: Date },
}, { timestamps: true });

export default mongoose.model<ISubscription>('Subscription', SubscriptionSchema);

File: backend/app/models/subscription.py
from datetime import datetime, timezone
from typing import Optional, Literal
from beanie import Document, PydanticObjectId
from pydantic import Field


class Subscription(Document):
    client_id: PydanticObjectId = Field(alias="clientId")
    plan_id: PydanticObjectId = Field(alias="planId")
    owner_id: PydanticObjectId = Field(alias="ownerId")
    status: Literal[
        "PENDING_ACTIVATION", "ACTIVE", "PAUSED", "CANCELLED", "EXPIRED", "FAILED"
    ] = "PENDING_ACTIVATION"
    start_date: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc), alias="startDate"
    )
    next_billing_date: datetime = Field(alias="nextBillingDate")
    notes: Optional[str] = None
    payment_failure_count: int = Field(default=0, alias="paymentFailureCount")
    last_payment_attempt: Optional[datetime] = Field(
        default=None, alias="lastPaymentAttempt"
    )
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    class Settings:
        name = "subscriptions"
        indexes = [
            "owner_id",
            "client_id",
            "status",
            "next_billing_date",
        ]
,
            "status",
            "next_billing_date",
        ]

    class Config:
        json_encoders = {
            PydanticObjectId: str
        }
